### PR TITLE
Fix `Handler.eval_step` and make `Logic` and `Handler` typed

### PR DESCRIPTION
**Pull request by asi1024**

### CODE REVIEW

**Comment by asi1024** [comment]
Resolved conflicts. PTAL.

File: pytorch_pfn_extras/handler.py [hunk 1]
```diff
@@ -391,7 +771,7 @@ def eval_step(self, evaluator, batch_idx, batch, complete_fn):
                 training step.
         """
         for _, sm, rt in self._runtime_iterator(evaluator.models):
-            rt.eval_pre_step(sm, evaluator, batch_idx, batch)
+            rt.eval_pre_step(evaluator, sm, batch_idx, batch)
```
Comment (asi1024) [inline pytorch_pfn_extras/handler.py:774]: note: bug fix.

**Comment by kmaehashi** [comment]
Could you resolve conflicts?

**Comment by asi1024** [comment]
I will resolve conflicts after #330 is merged.

**Comment by kmaehashi** [comment]
Merged #330.

**Pull request merged**